### DISCUSSION TITLE
`hybrid_property.inplace.expression` that would return `composite` column or tuple

### DISCUSSION THREAD

**Discussion by candiduslynx** [sqlalchemy/sqlalchemy#11756] [Usage Questions] [answered]
Hello!

I have a use case for the `hybrid_property.inplace.expression` that should return a `composite` column or a tuple.

<details><summary>Code snippet</summary>
<p>

```python
from collections.abc import Sequence
from dataclasses import dataclass
from datetime import datetime
from enum import StrEnum
from typing import Self

from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import (
    DeclarativeBase,
    InstrumentedAttribute,
    Mapped,
    aliased,
    composite,
    foreign,
    mapped_column,
    relationship,
)
from sqlalchemy.sql import Select, and_, literal, select


class Mode(StrEnum):
    AUTOMATIC = "AUTOMATIC"
    MANUAL = "MANUAL"


class Status(StrEnum):
    VERIFIED = "VERIFIED"
    FAILED = "FAILED"
    UNKNOWN = "UNKNOWN"


@dataclass(frozen=True, kw_only=False, order=True)
class Validation:
    mode: Mode
    status: Status


class Event(DeclarativeBase):
    parent_id: Mapped[str] = mapped_column(primary_key=True)
    at: Mapped[datetime] = mapped_column(primary_key=True)  # event's own versioning

    mode: Mode = mapped_column()
    status: Status = mapped_column()
    validation: Mapped[Validation] = composite(mode, status)

    _latest: type[Self] | None = None
    @classmethod
    def latest(cls) -> type["Event"]:
        if cls._latest is None:
            cls._latest = aliased(
                cls,
                select(cls)
                .distinct(
                    cls.parent_id,
                )
                .order_by(
                    cls.parent_id,
                    cls.at.desc(),
                )
                .cte("latest_event"),
                )
        return cls._latest

    @staticmethod
    def current_validation(events: Sequence["Event"]) -> Validation:
        manual, auto = (
            next(
                (
                    e.validation
                    for e in events
                    if e.mode == mode
                ),
                None,
            )
            for mode in (Mode.MANUAL, Mode.AUTOMATIC)
        )
        if manual is not None:
            return manual
        if auto is not None:
            return auto
        return Validation(mode=Mode.AUTOMATIC, status=Status.UNKNOWN)


# See `_choose_validation_status`; this logic should match!
def _compute_validation(
    parent_id: InstrumentedAttribute[str],
) -> Validation:
    def _select(mode: Mode) -> Select[tuple[Validation]]:
        return select(Event.latest().validation).where(
            and_(
                Event.latest().parent_id == parent_id,
                Event.latest().mode == mode,
            ),
        )

    return (
        _select(Mode.MANUAL)
        .union_all(
            _select(Mode.AUTOMATIC),
            select(
                literal(Mode.AUTOMATIC),
                literal(Status.UNKNOWN),
            ),
        )
        .limit(1)
    )


class Parent(DeclarativeBase):
    id: Mapped[str] = mapped_column(primary_key=True)
    at: Mapped[datetime] = mapped_column(primary_key=True) # parent's own versioning

    events: Mapped[Sequence[Event]] = relationship(
        Event.latest,
        primaryjoin=lambda: foreign(Event.latest().parent_id) == Parent.id,
        # sqlalchemy.orm.exc.UnmappedColumnError: Can't execute sync rule for destination column
        # XXX; mapper YYY does not map this column.
        # Try using an explicit `foreign_keys` collection which does not include this column (or use
        # a viewonly=True relation).
        viewonly=True,
        # Setting this is somewhat pointless because of the above but is necessary if we wish to use
        # the relationship on a newly constructed object.
        init=False,
        collection_class=list,
        lazy="raise",
    )

    @hybrid_property
    def validation(self) -> Validation:
        return Event.current_validation(self.events)

    @validation.inplace.expression
    @classmethod
    def _validation(cls):
        return _compute_validation(cls.id)
```

</p>
</details> 

`_compute_validation` is my latest attempt on this but I still fail to wrap my head around the proper signature here.

Additionally, the relationship is constructed this way as sometimes I need to check validation before the object is actually flushed to the database.

If it would be of any help, the engine currently is PostgreSQL.

**Comment by zzzeek** [comment]
there's a lot going on here but the first part that might not work is this:

```
 select(events.validation)
```

i dont know that we support a SELECT of a composite like that, it's not going to expand in the way you want.   you need to build that up manually, like

```py
select(tuple_(Event.mode, Event.status))
```

see if that SELECT works by itself first

**Reply by candiduslynx** [reply]
Oh, you've probably read the obsolete example. I've corrected it to include the current iteration with `select(Event.latest().validation)`.

For now the last attempt coerces the return as:
```python
q = (
        _select(Mode.MANUAL)
        .union_all(
            _select(Mode.AUTOMATIC),
            select(
                literal(Mode.AUTOMATIC),
                literal(Status.UNKNOWN),
            ),
        )
        .limit(1)
    )
    return Validation(q.selected_columns["mode"], q.selected_columns["status"])
```

**Reply by candiduslynx** [reply]
BTW, `select(tuple_(Event.mode, Event.status))` is not required at all, the `select(x.validation)` works fine (but it actually selects the columns instead of a tuple.

**Reply by zzzeek** [reply, marked as answer]
OK, yes a straight `select(Cls.composite)` these days will give you the composite class.

However the way the `@expression` is written here is not how a hybrid property works.    the expression element returns a SQL element and is only rendered as a SQL statement that's sent to the database.  it does not return a Validation object, there's no data to be populated into the `Validation` object you're building there.   also `.expression` is not intended to return a SELECT statement, it returns only a column expression.     

so you could have this column expression return `tuple_()`, but then use [typedecorator](https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator) to convert the tuple into a Validation object.   that's because you're on PostgreSQL which has full support of tuple SQL types.

alternatively, the ORM packaging for this composite class thing is kind of internal and it's based on [Bundle](https://docs.sqlalchemy.org/en/20/orm/queryguide/select.html#bundles).    there's no ready-to-go API to make ad-hoc composite expressions.   To make it do everything, like comparison `MyObject.validation == other_object.validation` as well as returning objects, you'd need to [emulate the approach that Composite uses](https://github.com/sqlalchemy/sqlalchemy/blob/98836f07f3219ab49aaabeb3a647d9a9799aacd5/lib/sqlalchemy/orm/descriptor_props.py#L708-L848) so that you have row processor as well as comparator support.   

This is a heavier lift so below is a fully working example that's based on the composite docs at https://docs.sqlalchemy.org/en/20/orm/composites.html , combining [hybrid custom comparators](https://docs.sqlalchemy.org/en/20/orm/extensions/hybrid.html#building-custom-comparators) with [custom bundle objects](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#sqlalchemy.orm.Bundle.create_row_processor), all public APIs, work with this example first as a means to see how to make a hybrid that delivers composites

```py
import dataclasses
from typing import Any
from typing import TYPE_CHECKING

from sqlalchemy import and_
from sqlalchemy import Any
from sqlalchemy import ColumnElement
from sqlalchemy import create_engine
from sqlalchemy import select
from sqlalchemy.ext.hybrid import Comparator
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import Bundle
from sqlalchemy.orm import composite
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session


@dataclasses.dataclass
class Point:
    x: int
    y: int

    def multiply(self, other):
        return Point(self.x * other.x, self.y * other.y)


class Base(DeclarativeBase):
    pass


class Vertex(Base):
    __tablename__ = "vertices"

    id: Mapped[int] = mapped_column(primary_key=True)

    if TYPE_CHECKING:
        x1: Mapped[int]
        x2: Mapped[int]
        y1: Mapped[int]
        y2: Mapped[int]

    start: Mapped[Point] = composite(mapped_column("x1"), mapped_column("y1"))
    end: Mapped[Point] = composite(mapped_column("x2"), mapped_column("y2"))

    @hybrid_property
    def start_times_end(self) -> Point:
        return self.start.multiply(self.end)

    @start_times_end.inplace.comparator
    @classmethod
    def _start_times_end(cls) -> PointComparator:  # type: ignore[override]
        return PointComparator(cls.x1 * cls.x2, cls.y1 * cls.y2)


class PointBundle(Bundle[Point]):
    """docs:

    https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#sqlalchemy.orm.Bundle.create_row_processor

    """
    def create_row_processor(self, query, procs, labels):
        def proc(row):
            return Point(procs[0](row), procs[1](row))

        return proc


class PointComparator(Comparator[str]):
    """docs:

    https://docs.sqlalchemy.org/en/20/orm/extensions/hybrid.html#building-custom-comparators

    """
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __clause_element__(self) -> PointBundle:
        return PointBundle("mybundle", self.x, self.y)

    def __eq__(self, other: Any) -> ColumnElement[bool]:
        return and_(self.x == other.x, self.y == other.y)

#
#    TODO:  other comparisons
#
#    def __ne__(self, other: Any) -> ColumnElement[bool]:
#        ...
#
#    def __lt__(self, other: Any) -> ColumnElement[bool]:
#        ...
#
#    def __gt__(self, other: Any) -> ColumnElement[bool]:
#        ...
#
#    def __le__(self, other: Any) -> ColumnElement[bool]:
#        ...
#
#    def __ge__(self, other: Any) -> ColumnElement[bool]:
#        ...


stmt = select(Vertex.start)

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

with Session(e) as sess:
    sess.add(Vertex(start=Point(3, 4), end=Point(5, 6)))
    sess.commit()

# select point by itself
with Session(e) as sess:
    point = sess.scalar(select(Vertex.start_times_end))
    print(point)


# comparison
with Session(e) as sess:
    point = sess.scalar(
        select(Vertex.start_times_end).where(
            Vertex.start_times_end == Point(15, 24)
        )
    )
    print(point)

```

**Reply by candiduslynx** [reply]
Thanks you!
I'll take a look and report on the results.

**Reply by candiduslynx** [reply]
So, returning with the results:

- Using type decorator works, but it's then quite impossible to construct queries that would check only a part of the composite (like, verify the `status` value only), so had to use comparator
- Comparator approach works, but the produced query for the `Parent` doesn't include those columns & generates the expression for every place the condition on the hybrid property is used

I think it would be great if the `select(Parent, Parent.validation)` would render a proper select instead of `select parent, (expression for the validation)` (as the parent, expression is cartesian product). It's also an issue when using some advanced filters as well (have to resort to rewriting queries as `where(exists().where(...))`.

**Reply by zzzeek** [reply]
>Comparator approach works, but the produced query for the Parent doesn't include those columns & generates the expression for every place the condition on the hybrid property is used

why *wouldnt* the expression be generated where the hybrid property is used?  are you looking for like "SELECT <expr> AS label ORDER BY label" kind of thing?  we support that if you use label() and then refer to that string by name in order by, see https://docs.sqlalchemy.org/en/20/tutorial/data_select.html#ordering-or-grouping-by-a-label 

> I think it would be great if the select(Parent, Parent.validation) would render a proper select instead of select parent, (expression for the validation) (as the parent, expression is cartesian product).

not following.   Using my example, if I do this:

     stmt = select(Vertex, Vertex.start_times_end)
     print(stmt)

I get exactly the SQL we'd expect:

```py
SELECT vertices.id, vertices.x1, vertices.y1, vertices.x2, vertices.y2, vertices.x1 * vertices.x2 AS anon_1, vertices.y1 * vertices.y2 AS anon_2 
FROM vertices
```

there's only one FROM clause so no cartesian product

**Reply by candiduslynx** [reply]
In your example you don't do joins, though.

What I'm looking forward to is this:
```
with subq1 as (...)
with parent as (... select from Parent + subq1, the included cols include status, reason, status1, reason1 (as we can have multiple modes attached to validations)
select ... from parent where parent.status1 in (...)
```

Now this will actually do the whole subquery for the `status1` column instead of reusing the available in the `parent` subquery

**Reply by zzzeek** [reply]
I would need a working example to try.  if your hybrid accesses other tables then it has to return a correlated subquery that's self-contained.  otherwise you're going outside of what hybrids are useful for.

**Reply by zzzeek** [reply]
a hybrid is not going to automatically add CTEs to the statement.  that's out of scope

**Reply by zzzeek** [reply]
you might want to look at [hybrid_method](https://docs.sqlalchemy.org/en/20/orm/extensions/hybrid.html#sqlalchemy.ext.hybrid.hybrid_method) where you can pass in the FROM elements which you want it to use in order to generate the column-oriented expression

**Comment by CaselIT** [comment]
> so you could have this column expression return tuple_(), but then use [typedecorator](https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator) to convert the tuple into a Validation object. that's because you're on PostgreSQL which has full support of tuple SQL types.

It's actually not really supported to return tuples: https://github.com/sqlalchemy/sqlalchemy/discussions/9951

also IIRC the compiler will error if it finds a tuple type in the list of select columns
